core/src/main/java/io/questdb/griffin/engine/functions/str/SplitPartFunctionFactory.java

```diff
@@ -0,0 +1,210 @@
+        private final Function strFunc;
+        private final Function delimiterFunc;
+        private final Function indexFunc;
+
+        public SplitPartFunction(Function strFunc, Function delimiterFunc, Function indexFunc) {
+            super(strFunc, delimiterFunc);
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+            this.indexFunc = indexFunc;
+        }
+
+        @Override
+        int getIndex(Record rec) {
+            return indexFunc.getInt(rec);
+        }
+
+        @Override
+        public Function getLeft() {
+            return strFunc;
+        }
+
+        @Override
+        public Function getCenter() {
+            return delimiterFunc;
+        }
+
+        @Override
+        public Function getRight() {
+            return indexFunc;
+        }
+    }
+
+    private static class SplitPartConstIndexFunction extends AbstractSplitPartFunction implements BinaryFunction {
+        private final Function strFunc;
+        private final Function delimiterFunc;
+        private final int index;
+
+        public SplitPartConstIndexFunction(Function strFunc, Function delimiterFunc, int index) {
+            super(strFunc, delimiterFunc);
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+            this.index = index;
+        }
+
+        @Override
+        int getIndex(Record rec) {
+            return index;
+        }
+
+        @Override
+        public Function getLeft() {
+            return strFunc;
+        }
+
+        @Override
+        public Function getRight() {
+            return delimiterFunc;
+        }
+    }
+
+    private static abstract class AbstractSplitPartFunction extends StrFunction {
+        private final StringSink sink = new StringSink();
+        private final StringSink sinkB = new StringSink();
+
+        private final Function strFunc;
+        private final Function delimiterFunc;
+
+        public AbstractSplitPartFunction(Function strFunc, Function delimiterFunc) {
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+        }
+
+        abstract int getIndex(Record rec);
+
+        @Override
+        public CharSequence getStr(Record rec) {
+            return getStr0(rec, sink);
+        }
+
+        @Override
+        public CharSequence getStrB(Record rec) {
+            return getStr0(rec, sinkB);
+        }
+
+        @Override
+        public void getStr(Record rec, CharSink sink) {
+            getStr0(rec, sink);
+        }
+
+        @Nullable
+        private <S extends CharSink> S getStr0(Record rec, S sink) {
+            var str = strFunc.getStr(rec);
+            var delimiter = delimiterFunc.getStr(rec);
+            var index = getIndex(rec);
+            if (str == null || delimiter == null || index == Numbers.INT_NaN) {
+                return null;
+            }
+            if (index == 0) {
+                return sink;
```

Comment:
Same consideration on the `field position must not be zero` error.

Reply:
How can I return an error from `getStr()`? The interface methods do not declare any thrown exceptions, and `io.questdb.griffin.SqlException` is a checked exception. Can I modify the `Function` interface methods?

Reply:
Sorry, forgot to mention that. `getStr()` and other data accessor methods can't throw. We'd have to require the index argument to be a constant or a runtime constant (think, a bind variable) and evaluate its value only a single time per constructor call (if it's a constant) or in the `void init(SymbolTableSource, SqlExecutionContext)` call (if it's a runtime constant). See how it's done in `MatchStrFunctionFactory`, as an example.

Reply:
Done. I've moved this to the `ArgFunctions` utility class that I've created, to init all runtime constant and dynamic arg functions, and to validate all runtime constant arg functions.